File: src/android/app/src/main/java/info/cemu/Cemu/inputoverlay/RoundButton.java
package info.cemu.Cemu.inputoverlay;

import android.content.res.Resources;
import android.graphics.Rect;

import androidx.annotation.DrawableRes;

public class RoundButton extends Button {
    int centreX;
    int centreY;
    int radius;

    RoundButton(Resources resources, @DrawableRes int pressedButtonId, @DrawableRes int notPressedButtonId, ButtonStateChangeListener buttonStateChangeListener, InputOverlaySurfaceView.OverlayButton overlayButton, InputOverlaySettingsProvider.InputOverlaySettings settings) {
        super(resources, pressedButtonId, notPressedButtonId, buttonStateChangeListener, overlayButton, settings);
        configure();
    }

    @Override
    protected void configure() {
        var rect = settings.getRect();
        centreX = rect.centerX();
        centreY = rect.centerY();
        radius = Math.min(rect.width(), rect.height()) / 2;
        var iconRect = new Rect(
                centreX - radius,
                centreY - radius,
                centreX + radius,
                centreY + radius);
        iconPressed.setBounds(iconRect);
        iconNotPressed.setBounds(iconRect);
    }

    @Override
    public boolean isInside(int x, int y) {
        return ((x - centreX) * (x - centreX) + (y - centreY) * (y - centreY)) <= radius * radius;
    }

}

File: src/android/app/src/main/java/info/cemu/cemu/emulation/inputoverlay/inputs/RoundButton.kt
package info.cemu.cemu.emulation.inputoverlay.inputs

import android.graphics.Canvas
import android.graphics.Rect
import info.cemu.cemu.emulation.inputoverlay.fillCircleWithStroke
import info.cemu.cemu.emulation.inputoverlay.inputs.innerdrawing.ButtonInnerDrawing
import kotlin.math.min

class RoundButton(
    private val innerDrawing: ButtonInnerDrawing,
    onButtonStateChange: (state: Boolean) -> Unit,
    private val alpha: Int,
    rect: Rect,
) : Button(onButtonStateChange, rect) {
    private var centerX: Float = 0f
    private var centerY: Float = 0f
    private var radius: Float = 0f
    private var radius2: Float = 0f

    init {
        configure()
    }

    override fun configure() {
        centerX = rect.centerX().toFloat()
        centerY = rect.centerY().toFloat()
        radius = min(rect.width().toFloat(), rect.height().toFloat()) / 2f
        radius2 = radius * radius
        innerDrawing.configure(rect, alpha)
        configureColors(alpha)
    }

    override fun drawInput(canvas: Canvas) {
        val fillColor: Int
        val strokeColor: Int
        if (state) {
            fillColor = activeFillColor
            strokeColor = activeStrokeColor
        } else {
            fillColor = inactiveFillColor
            strokeColor = inactiveStrokeColor
        }
        canvas.fillCircleWithStroke(centerX, centerY, radius, paint, fillColor, strokeColor)
        innerDrawing.draw(canvas, state)
    }

    override fun isInside(x: Float, y: Float): Boolean {
        val dx = x - centerX
        val dy = y - centerY
        return dx * dx + dy * dy <= radius2
    }
}
